plasmapy/atomic/classes.py

```diff
@@ -144,109 +151,145 @@ def __init__(self,
 
     @property
-    def ion(self):
+    def ion(self) -> str:
+        r"""Returns the ion symbol, or raises an InvalidIonError if the
+        particle is not an ion."""
+        if not self._ion_symbol:
+            raise InvalidIonError(self._ion_errmsg)
         return self._ion_symbol
 
     @property
-    def atomic_number(self):
-        if self._atomic_number is not None:
-            return self._atomic_number
-        else:
-            raise InvalidElementError(
-                f"Particle '{self._particle_symbol}' is not an element, "
-                f"so no atomic number is available.")
+    def element_name(self) -> str:
+        r"""Returns the name of the element, or raises an InvalidElementError
+        if the particle is not an element, isotope, or ion."""
+        if not self.element:
+            raise InvalidElementError(self._element_errmsg)
+        return self._element_name
 
     @property
-    def mass_number(self):
-        if self._mass_number is not None:
-            return self._mass_number
-        else:
-            raise InvalidIsotopeError(
-                f"Particle '{self._particle_symbol}' is not an isotope, so "
-                f"no mass number is available.")
+    def atomic_number(self) -> int:
+        if not self.element:
+            raise InvalidElementError(self._element_errmsg)
+        return self._atomic_number
 
     @property
-    def lepton_number(self):
-        if self._lepton_number is not None:
-            return self._lepton_number
-        else:
-            raise AtomicError(f"The leption number for "
-                              f"'{self._particle_symbol}' is not available.")
+    def mass_number(self) -> int:
+        if not self.isotope:
+            raise InvalidIsotopeError(self._isotope_errmsg)
+        return self._mass_number
 
     @property
-    def baryon_number(self):
-        if self._baryon_number is not None:
-            return self._baryon_number
-        elif self._mass_number is not None:
-            return self._mass_number
-        else:
-            raise AtomicError("The baryon number for "
-                              f"'{self._particle_symbol}' is not available.")
+    def lepton_number(self) -> int:
+        if self._lepton_number is None:
+            raise AtomicError(
+                f"The lepton number for {self.particle} is not available.")
+        return self._lepton_number
 
     @property
-    def half_life(self):
-        if self._half_life is not None:
-            return self._half_life
-        else:
+    def baryon_number(self) -> int:
+        r"""Returns the baryon number, or raises a MissingAtomicDataError if
+        the baryon number is unavailable."""
+        if self._baryon_number is None:
+            raise AtomicError(
+                f"The baryon number for '{self.particle}' is not "
+                f"available.")
+        return self._baryon_number
+
+    @property
+    def half_life(self) -> u.Quantity:
```

Comment:
I kind of wonder whether we couldn't do `-> 1/u.s` here. You know, to signalize that this returns an inverse of time.

I think @units.quantity_input also plays well with that in most cases (*grumble grumble electronvolts*), but I'll have to check back. I think I should have some time to play around with unit checks and annotations next week, I feel like it's sort of finishing gestating in my head.

Reply:
Interesting point.  In terms of physics, it does make a lot of sense to use inverse seconds as a more natural unit.  The phrase "half-life" implies a timescale and so from that name I think most people would expect seconds.  I tried searching for "inverse of half-life" since I didn't know the name of it, and the first term I found was "radioactive decay constant", though that's less widely used than "half-life".  Do we think it would be worth having another attribute called `decay_constant` with attributes of inverse seconds?

Reply:
Oh crudddd yeah I really wasn't thinking here. Obviously it returns in seconds, haha! ^^'

What I meant was that we could change the return type hint from `-> u.Quantity`, as that's really generic, to `-> u.s`, as that's a `Quantity` and it's also more specific and informative.